### PR TITLE
Add support for Windows CA certificate store

### LINKED ISSUE #18020: Windows-ROOT as truststore
Is there a way to set Windows-ROOT (or any Windows Certificate Store in general) as truststore in openssl?
I think java has some similar feature.

Thanks in advance.

### PR DESCRIPTION
**Pull request by hlandau**
This PR adds support for the Windows CA certificate store.  It does this by introducing a new store in the default provider which recognises the `capi://` URI scheme. The URI used is just `capi://`, there are no arguments. This new store also becomes the default store when `capistore` is enabled at compile time (i.e., on Windows).

A build option is added (`capistore`) which can be disabled if this functionality is not desired.

This fixes #18020 but also implements my proposal in #18068 and is built on top of that fix. I can split these out into separate PRs if desired, but the former would preferably be merged before the latter.

### CODE REVIEW

**Comment by hlandau** [comment]
Manpage.

**Comment by petrovr** [comment]
I'm not convinced that proposed solution should change existing methods.
Implementation of Store2 API must be enough.

**Comment by petrovr** [comment]
About implementation - it is good to see a provider based sample.  I have a number of engine based implementation including own by_store x509 lookup for openssl before 3.
Also I expect not so generic function names, i.e. with store as part of name. Perhaps current names are acceptable as static functions.
I could not see "error", "expect" and "find" methods. May be "expect" is superseded by "set...param...". If think that other two are required.

**Comment by petrovr** [comment]
P.S.I think that store name could be extracted from URI.

File: include/internal/common.h [hunk 1]
```diff
@@ -112,4 +116,15 @@ static ossl_inline int ossl_is_absolute_path(const char *path)
...
 
+static ossl_inline int ossl_is_uri(const char *s)
+{
+    const char *x;
+    for (x=s; ossl_isalnum(*x); ++x);
+#ifdef _WIN32
+    if (x-s <= 1)
+        return 0;
+#endif
+    return x > s && HAS_PREFIX(x, "://");
```
Comment (t8m) [inline include/internal/common.h:132]: This is certainly not right as URIs do not have to have the `//`

Comment (hlandau) [reply include/internal/common.h:132]: That's true. Basically my concern here is supporting compatibility with anyone using multiple relative paths in `SSL_CERT_DIR`.

If you agree it's rare enough not to be worth worrying about, I'll change it.

**Comment by kaduk** [comment]
Please register any new URI schemes with IANA: https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml -- provisional registrations are given on a "first come first served" policy.

**Comment by hlandau** [comment]
@beldmit It uses the system root CA store by default. Since this code uses an URI to specify the Windows store there is the possibility we could allow other stores to be specified.

**Comment by beldmit** [comment]
How do we override this default URI in run-time?

**Comment by hlandau** [comment]
The same way the URI is currently specified, e.g. via an environment variable, see this PR. I could add support for specifying different stores.

**Comment by beldmit** [comment]
It looks better to me if it is overrideable via API.

**Comment by hlandau** [comment]
That is also feasible.

**Comment by levitte** [comment]
As mentioned here, `capi:` isn't a good scheme name: https://github.com/openssl/openssl/issues/18068#issuecomment-1094849701

File: crypto/x509/by_store.c [hunk 2]
```diff
@@ -111,11 +111,24 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
...
     case X509_L_ADD_STORE:
-        /* If no URI is given, use the default cert dir as default URI */
+        /* First try the newer default cert URI envvar. */
         if (argp == NULL)
+            argp = ossl_safe_getenv(X509_get_default_cert_uri_env());
+
+        /* If not set, see if we have a URI in the older cert dir envvar. */
+        if (argp == NULL) {
             argp = ossl_safe_getenv(X509_get_default_cert_dir_env());
+            if (argp != NULL && !ossl_is_uri(argp))
```
Comment (t8m) [inline crypto/x509/by_store.c:121]: If we keep the check for `://` in ossl_is_uri IMO we should NOT do this check here as I said that URIs are definitely valid without `//`.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/x509/by_store.c (hunk 2 above)
Comment (hlandau) [reply crypto/x509/by_store.c:121]: What currently happens in 3.0 is:

Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`, search store `foo:bar`
Specify `SSL_CERT_DIR=foo://bar` -> search dir `foo`, search dir `//bar`, search store `foo://bar`

So the question is how to fix this in a compatible way. Currently if someone specifies a store URI of `foo://bar`, the directory `foo` in the current directory is added as a root certificate directory which is quite alarming.

The possible fixes I see are:

Resolution A: If "://" is found in `SSL_CERT_DIR`, treat as an URI, else, as a list of paths
- URIs of form `scheme://` can use SSL_CERT_DIR, but URIs of form `scheme:` must use SSL_CERT_URI
- Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
- Pro: Fixes the ambiguity, which has potential security implications, for existing users
- Con: May cause breakage to anyone using a custom store with `foo:bar` URIs, who will need to use `SSL_CERT_URI` instead

Resolution B: If ":" is found in `SSL_CERT_DIR`, treat as an URI, else, as a single path
- Specify `SSL_CERT_DIR=foo` -> search dir `foo`
- Specify `SSL_CERT_DIR=foo:bar` -> search store `foo:bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
- Con: Very incompatible

Resolution C: keep the current behaviour for the existing environment variables and introduce two new environment variables
- Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar` store `foo:bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search dir `foo`, search dir `//bar` store `foo://bar`
- `SSL_CERT_DIR` is deprecated but continues to work as before. The potential security issues are documented.
- A new environment variable `SSL_CERT_URI` becomes the preferred way to specify a URI
- A new environment variable `SSL_CERT_PATH` becomes the preferred way to specify a list of paths. Unlike `SSL_CERT_DIR` it is never processed as a URI.

Comment (t8m) [reply crypto/x509/by_store.c:121]: I prefer C.

I'd accept a minor variant of A, partially combined with C:

 - URIs of form `scheme://` can use SSL_CERT_DIR, but URIs of form `scheme:` should use SSL_CERT_URI
 - Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`, search store `foo:bar`
 - Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
 - Pro: Partially fixes the ambiguity, which has potential security implications, for existing users
 - Con: Does not fix the ambiguity for `foo:bar` stores which means we would probably still need to deprecate SSL_CERT_DIR and introduce SSL_CERT_PATH

I am not sure the slight improvement of fixing the issue for existing users of `scheme://` with SSL_CERT_DIR is worth it.

Comment (levitte) [reply crypto/x509/by_store.c:121]: I'd prefer C as well (it resolves a naming bug as well.... Common nomenclature is that PATH indicates several concatenated directories while DIR is a single directory)

**Review by levitte — APPROVED** [review]
Looking good at this point.

**Comment by levitte** [comment]
(it took me a moment to grasp the fact that this is loading certificates only...  :wink:)

File: providers/stores.inc [hunk 3]
```diff
@@ -12,3 +12,6 @@
 #endif
 
 STORE("file", "yes", ossl_file_store_functions)
+#ifndef OPENSSL_NO_WINSTORE
+STORE("winstore", "no", ossl_winstore_store_functions)
```
Comment (t8m) [inline providers/stores.inc:16]: The "no" means fips=no - imo there is no reason to have it so as there is no difference between file and winstore stores in this regard.

**Review by t8m — CHANGES_REQUESTED** [review]
Minor nits

**Comment by hlandau** [comment]
Updated. @t8m

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@levitte still OK?

**Comment by levitte** [comment]
Sorry, didn't see this before now.  Yes, still ok

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Given the conflict is just in CHANGES.md which is trivial to resolve, I do not think a rebase is needed, the conflict can be fixed when merging.

**Comment by mattcaswell** [comment]
Has a request been made to register the URI scheme as per the [comment](https://github.com/openssl/openssl/pull/18070#issuecomment-1095709036) above?

**Comment by kaduk** [comment]
> Has a request been made to register the URI scheme as per the [comment](https://github.com/openssl/openssl/pull/18070#issuecomment-1095709036) above?

The combination of https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml and https://mailarchive.ietf.org/arch/browse/uri-review/ indicate "no"...

**Comment by hlandau** [comment]
Rebased.

@mattcaswell It hasn't. It feels overkill. What do we want to do? URI registration is specification required. If we want to register it, I'll need to write a spec and push it through the process.

**Comment by levitte** [comment]
> Rebased.
> 
> @mattcaswell It hasn't. It feels overkill, at least at this stage. What do we want to do? URI registration is specification required. If we want to register it, I'll need to write a spec and push it through the process.
> 
> Wondering what other OpenSSL store providers do.

I'm only aware of PKCS#11, for which there already is a well defined URI scheme: https://datatracker.ietf.org/doc/html/rfc7512

The question to ask is, what if someone else comes up with a `winstore://` URI scheme, independent of this one?  Will their scheme work the same as ours, or will people have to know which library it's sent through for proper use?  After all, "winstore" is a fairly obvious name that anyone could come up with...

It can be argued, though, that we don't have to have that registration in place (even provisional) before the master branch becomes a next release.  That gives us a time frame for doing this.

Also, I might point out that it doesn't necessarily fall on *you* to write the URI spec.  Someone else can do that just as well.

**Comment by hlandau** [comment]
Of course, nothing even comes after `winstore://`. In future we could maybe support named stores since the wincrypt API does have such a feature. `winstore://specific-root-store-name`. But this might be so rarely used so as to not be worth it.

Writing a spec wouldn't take too long, though I wonder how useful this URI scheme is going to be to anyone else. I guess it's conceivably possible other parties using URIs to represent certificate stores might find it useful. So maybe it is worth doing. Or we could go with a scheme name like `vnd-openssl-winstore`. There was an abortive effort to standardise such a scheme: https://lists.w3.org/Archives/Public/uri/2002Dec/0006.html https://datatracker.ietf.org/doc/html/draft-king-vnd-urlscheme-03

**Comment by levitte** [comment]
> There was an abortive effort to standardise such a scheme: https://lists.w3.org/Archives/Public/uri/2002Dec/0006.html https://datatracker.ietf.org/doc/html/draft-king-vnd-urlscheme-03

That is, of course, that possibility too.

Too bad, BTW, that no one pointed at (or knew about) that draft when we made this:

https://github.com/openssl/openssl/blob/a6843e6ae8ae0551aae8555783f06dab7951f112/apps/include/engine_loader.h#L15-L16

**Comment by mattcaswell** [comment]
> I'll need to write a spec and push it through the process.

My understanding is that it is sufficient to just create and register a draft spec. It doesn't necessarily need to be pushed all the way through the process (i.e. it can just remain as draft and then expire). So the spec doesn't even need to be that good!

@kaduk can probably advise on how to do it

**Comment by hlandau** [comment]
Yeah, if we decide to get it registered, I will just make a PR with a design document for the URI scheme, which should be enough.

@levitte I actually was about to suggest `org.openssl.winstore://` before I found that `vnd-` proposal. Since that proposal was never finalised, and we apparently have a past history of using `org.openssl.engine://`, I suggest we go with `org.openssl.winstore://` for consistency with ourselves.

**Comment by levitte** [comment]
> @levitte I actually was about to suggest `org.openssl.winstore://` before I found that `vnd-` proposal. Since that proposal was never finalised, and we apparently have a past history of using `org.openssl.engine://`, I suggest we go with `org.openssl.winstore://` for consistency with ourselves.

To be noted: I pushed for `org.openssl.engine://` mainly because engines are on a death path, so that URI scheme would be foreseeably temporary.  Not sure if that impacts how we think about winstore, though, but I thought it was worth mentioning...

**Comment by hlandau** [comment]
What's the review status on this?

**Comment by hlandau** [comment]
Rebased

**Comment by mattcaswell** [comment]
> Rebased

Looks like it needs another one.

> What's the review status on this?

Isn't it stuck on registering the URI scheme?

**Comment by mattcaswell** [comment]
I prefer (1) followed by (2). I don't this we should do (3).

**Comment by t8m** [comment]
We can even do 2 now and 1 later.

**Comment by hlandau** [comment]
Rebased, and updated for now to use `org.openssl.winstore`, consistent with our existing usage of `org.openssl.engine`, etc. We could merge this now and consider accepting aliasing `winstore` when we can get a URI scheme registered.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@levitte still OK?

File: CHANGES.md [hunk 4]
```diff
@@ -179,6 +179,27 @@ OpenSSL 3.1
...
+   list of paths or an URI, which is ambiguous. Setting `SSL_CERT_PATH` causes
+   `SSL_CERT_DIR` to be ignored for the purposes of determining root certificate
+   directories, and setting `SSL_CERT_URI` causes `SSL_CERT_DIR` to be ignored
+   for the purposes of determining root certificate stores.
+
+   *Hugo Landau*
+
+ * Support for loading root certificates from the Windows certificate store
+   has been added. The support is in the form of a store which recognises
+   the URI string of `winstore://`. This store is enabled by default and can
```
Comment (levitte) [inline CHANGES.md:198]: This isn't quite right for the moment...

**Review by levitte — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
Other than the nit in CHANGES.md, this looks good enough for me.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Pull request closed without merging**